homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,432 @@
+        self._async_activation_listeners: Dict[str, Callable] = {}
+        self._async_deactivation_listeners: Dict[str, str] = {}
+        self._hass: "HomeAssistant" = hass
+        self._latest_instances: List[ScheduleInstance] = []
+        self.schedules: Dict[str, dict] = {}
+
+    @callback
+    def as_dict(self, schedule_id: str) -> dict:
+        """Return a schedule in dict form."""
+        schedule = self.schedules[schedule_id]
+        return {
+            CONF_SCHEDULE_ID: schedule_id,
+            CONF_ENTITY_ID: schedule[CONF_ENTITY_ID],
+            CONF_START_DATETIME: schedule[CONF_START_DATETIME].isoformat(),
+            CONF_DURATION: schedule[CONF_DURATION],
+            CONF_RRULE: str(schedule[CONF_RRULE]),
+            CONF_ACTIVATION_CONTEXT_ID: schedule[CONF_ACTIVATION_CONTEXT].id,
+        }
+
+    @callback
+    def async_create(
+        self,
+        entity_id: str,
+        start_datetime: datetime,
+        *,
+        duration: Optional[int] = None,
+        rrule_str: Optional[str] = None,
+        activation_context_id: Optional[str] = None,
+    ) -> dict:
+        """Create a schedule.
+
+        The rrule_str parameter must be an RFC5545-compliant string.
+        The duration parameter is a number of seconds a schedule instance should last.
+        """
+        if start_datetime < datetime.now():
+            raise ValueError("Dates in the past are not allowed")
+
+        schedule_id = uuid4().hex
+        schedule = {
+            CONF_ENTITY_ID: entity_id,
+            CONF_START_DATETIME: start_datetime,
+            CONF_DURATION: duration,
+            CONF_RRULE: rrulestr(rrule_str) if rrule_str else None,
+            CONF_ACTIVATION_CONTEXT: Context(id=activation_context_id)
+            if activation_context_id
+            else Context(),
+        }
+        self.schedules[schedule_id] = schedule
+
+        self.async_save()
+        _LOGGER.info('Created schedule "%s": %s', schedule_id, schedule)
+
+        self.async_schedule_next_instance(schedule_id)
+
+        return self.as_dict(schedule_id)
+
+    @callback
+    def async_delete(
+        self, schedule_id: str, *, remove_listeners: bool = True, save: bool = True
+    ) -> dict:
+        """Delete a schedule."""
+        if schedule_id not in self.schedules:
+            raise KeyError
+
+        # Since there are cases where listeners might not exist (like deleting a
+        # schedule after it completes), we give the caller the ability to delete the
+        # schedule without attempting to cancel nonexistent listeners:
+        if remove_listeners:
+            self.async_delete_latest_instance(schedule_id)
+
+        return_payload = self.as_dict(schedule_id)
+        self.schedules.pop(schedule_id)
+
+        if save:
+            self.async_save()
+        _LOGGER.info('Deleted schedule "%s"', schedule_id)
+
+        return return_payload
+
+    async def async_delete_latest_instance(self, schedule_id: str) -> None:
```

Comment:
cancel is a better description. When we cancel something, are we also undo-ing it ?

Reply:
Will change. RIght now, yes, canceling also unwinds the scene.